src/function/algorithm/wekaly_connected_component.cpp

```diff
@@ -0,0 +1,116 @@
+#include "function/gds/gds_function_collection.h"
+#include "function/gds_function.h"
+#include "graph/graph.h"
+#include "main/client_context.h"
+#include "processor/result/factorized_table.h"
+
+using namespace kuzu::binder;
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+
+namespace kuzu {
+namespace function {
+
+class WeaklyConnectedComponentLocalState : public GDSLocalState {
+public:
+    explicit WeaklyConnectedComponentLocalState(main::ClientContext* context) {
+        auto mm = context->getMemoryManager();
+        nodeIDVector = std::make_unique<ValueVector>(*LogicalType::INTERNAL_ID(), mm);
+        groupVector = std::make_unique<ValueVector>(*LogicalType::INT64(), mm);
+        nodeIDVector->state = DataChunkState::getSingleValueDataChunkState();
+        groupVector->state = DataChunkState::getSingleValueDataChunkState();
+        vectors.push_back(nodeIDVector.get());
+        vectors.push_back(groupVector.get());
+    }
+
+    void materialize(graph::Graph* graph, const std::vector<int64_t>& groupArray,
+        FactorizedTable& table) const {
+        for (auto offset = 0u; offset < graph->getNumNodes(); ++offset) {
+            nodeIDVector->setValue<nodeID_t>(0, {offset, graph->getNodeTableID()});
+            groupVector->setValue<int64_t>(0, groupArray[offset]);
+            table.append(vectors);
+        }
+    }
+
+private:
+    std::unique_ptr<ValueVector> nodeIDVector;
+    std::unique_ptr<ValueVector> groupVector;
+    std::vector<ValueVector*> vectors;
+};
+
+class WeaklyConnectedComponent final : public GDSAlgorithm {
+public:
+    WeaklyConnectedComponent() = default;
+    WeaklyConnectedComponent(const WeaklyConnectedComponent& other) : GDSAlgorithm{other} {}
+
+    std::vector<common::LogicalTypeID> getParamTypeIDs() const override {
+        return std::vector<LogicalTypeID>{LogicalTypeID::ANY};
+    }
+
+    std::vector<std::string> getResultColumnNames() const override {
+        return {"node_id", "group_id"};
+    }
+
+    std::vector<common::LogicalType> getResultColumnTypes() const override {
+        return {*LogicalType::INTERNAL_ID(), *LogicalType::INT64()};
+    }
+
+    void initLocalState(main::ClientContext* context) override {
+        localState = std::make_unique<WeaklyConnectedComponentLocalState>(context);
+    }
+
+    void exec() override {
+        auto wccLocalState = localState->ptrCast<WeaklyConnectedComponentLocalState>();
```

Comment:
Can these initialization steps be done inside initLocalState? That looks exactly like the purpose of the `initLocalState` function.

Reply:
This is a casting not initialization.